### ISSUE TITLE
ACK storm in initial packets

### ISSUE THREAD

**Issue by kenmcmil** [private-octopus/picoquic#548]
Here is a log where the picoquic client seems to be sending repeated ACKs to ACK-only initial packets.  For example, client initial packet 2 is an ack to ack-only packet 4, and client packet 3 is an ack to ack-only packet 36.  The tester flagged an error at this point, as the number of ack-only packets sent by the client is greater than the number of non-ack-only packets it received from the server. 

```
Starting PicoQUIC connection to server IP = localhost, port = 4443
No root crt list specified, certificate will not be verified.
0c5fce861fb13a9e: Sending packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <0c5fce861fb13a9e>, <86a99dafa9700ae6>, Seq: 0, pl: 1227
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Prepared 1207 bytes
0c5fce861fb13a9e:     Crypto HS frame, offset 0, length 305: 0100012d0303606b...
0c5fce861fb13a9e:     padding, 898 bytes

0c5fce861fb13a9e: Sending 1252 bytes to 127.0.0.1:4443 at T=0.000000 (5851ccc92fde0)
Select returns 1232, from length 16
Local address updated
0c5fce861fb13a9e: Receiving 1232 bytes from 127.0.0.1:4443 at T=0.113073 (5851ccc94b791)
0c5fce861fb13a9e: Receiving packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, <0000000000000002>, Seq: 0, pl: 1207
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Decrypted 1190 bytes
0c5fce861fb13a9e:     ACK (nb=0), 0
0c5fce861fb13a9e:     padding, 1185 bytes

Processed 1232 bytes, state = 7 (client_handshake_start), return 0

Select returns 1232, from length 16
0c5fce861fb13a9e: Receiving 1232 bytes from 127.0.0.1:4443 at T=0.218035 (5851ccc965193)
0c5fce861fb13a9e: Receiving packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, <0000000000000002>, Seq: 4, pl: 1207
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Decrypted 1190 bytes
0c5fce861fb13a9e:     Crypto HS frame, offset 0, length 155: 0200009703039a63...
0c5fce861fb13a9e:     padding, 1031 bytes

Processed 1232 bytes, state = 7 (client_handshake_start), return 0

0c5fce861fb13a9e: Sending packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <0000000000000002>, <86a99dafa9700ae6>, Seq: 1, pl: 27
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Prepared 7 bytes
0c5fce861fb13a9e:     ACK (nb=1), 4, 0

0c5fce861fb13a9e: Sending 52 bytes to 127.0.0.1:4443 at T=0.218203 (5851ccc96523b)
Select returns 1407, from length 16
0c5fce861fb13a9e: Receiving 1407 bytes from 127.0.0.1:4443 at T=0.312069 (5851ccc97c0e5)
0c5fce861fb13a9e: Receiving packet type: 4 (handshake), S0, Version ff000012,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, <0000000000000002>, Seq: 0, pl: 1383
0c5fce861fb13a9e:     Decrypted 1366 bytes
0c5fce861fb13a9e:     Crypto HS frame, offset 0, length 1346: 0800003d003b0000...
0c5fce861fb13a9e:     padding, 16 bytes

Processed 1407 bytes, state = 10 (client_almost_ready), return 0

Almost ready!

0c5fce861fb13a9e: Sending packet type: 4 (handshake), S0, Version ff000012,
0c5fce861fb13a9e:     <0000000000000002>, <86a99dafa9700ae6>, Seq: 0, pl: 80
0c5fce861fb13a9e:     Prepared 60 bytes
0c5fce861fb13a9e:     ACK (nb=0), 0
0c5fce861fb13a9e:     Crypto HS frame, offset 0, length 52: 14000030fc16da69...

0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S0,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 0 (0), Phi: 0,
0c5fce861fb13a9e:     Prepared 216 bytes
0c5fce861fb13a9e:     NEW CONNECTION ID[1]: 0xe1fdd8d5bf6ecb39, ae4de4bcec8a0438c8f4cf9c25376d1e
0c5fce861fb13a9e:     NEW CONNECTION ID[2]: 0x0522ee324c648303, 102d4f4fba4e299f516ade6cce61a5b5
0c5fce861fb13a9e:     NEW CONNECTION ID[3]: 0x0ba8d5a218f70d0b, aa71887658f70d6556efa2366a8aa045
0c5fce861fb13a9e:     NEW CONNECTION ID[4]: 0x38026e7774698253, fa9bcace5af252462dc6ba6203e0c67c
0c5fce861fb13a9e:     NEW CONNECTION ID[5]: 0xa2849b98330f73c7, 7f9aa3139736f2a236741a2a54e854b8
0c5fce861fb13a9e:     NEW CONNECTION ID[6]: 0x6edbeccf181b5362, 9567a4c0fd04182862e86599e9dbfbd2
0c5fce861fb13a9e:     NEW CONNECTION ID[7]: 0x547227128e671bc5, ad1b0b39bb79c496df6c7c92069930a3
0c5fce861fb13a9e:     NEW CONNECTION ID[8]: 0xd02d41b90bd79e1d, 8e188c83ade048ee108494ad70fa6f89

0c5fce861fb13a9e: Sending 349 bytes to 127.0.0.1:4443 at T=0.312149 (5851ccc97c135)
Received SNI: localhost
ALPN not received.
Received transport parameter TLS extension (51 bytes):
Version: ff000012
    Supported version (8 bytes):
        ff000012
        49565931
    Extension list (36 bytes):
        Extension type: 5, length 2 (0x0005 / 0x0002), 6000
        Extension type: 4, length 4 (0x0004 / 0x0004), 80004000
        Extension type: 1, length 1 (0x0001 / 0x0001), 3c
        Extension type: 6, length 2 (0x0006 / 0x0002), 6000
        Extension type: 7, length 2 (0x0007 / 0x0002), 6000
        Extension type: 8, length 1 (0x0008 / 0x0001), 10

Connection established. Version = ff000012, I-CID: c5fce861fb13a9e
Opening stream 0 to GET /index.html
0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S0,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 1 (1), Phi: 0,
0c5fce861fb13a9e:     Prepared 24 bytes
0c5fce861fb13a9e:     Stream 0, offset 0, length 17, fin = 1: 474554202f696e64...
0c5fce861fb13a9e:     padding, 4 bytes

0c5fce861fb13a9e: Sending 53 bytes to 127.0.0.1:4443 at T=0.312226 (5851ccc97c182)
0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S0,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 2 (2), Phi: 0,
0c5fce861fb13a9e:     Prepared 1411 bytes
0c5fce861fb13a9e:     ping, 1 bytes
0c5fce861fb13a9e:     padding, 1410 bytes

0c5fce861fb13a9e: Sending 1440 bytes to 127.0.0.1:4443 at T=0.312314 (5851ccc97c1da)
Select returns 1232, from length 16
0c5fce861fb13a9e: Receiving 1232 bytes from 127.0.0.1:4443 at T=0.434670 (5851ccc999fce)
0c5fce861fb13a9e: Receiving packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, <0000000000000002>, Seq: 18, pl: 1207
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Decrypted 1190 bytes
0c5fce861fb13a9e:     ACK (nb=0), 1
0c5fce861fb13a9e:     padding, 1185 bytes

Processed 1232 bytes, state = 13 (client_ready_start), return 0

0c5fce861fb13a9e: Sending packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <0000000000000002>, <86a99dafa9700ae6>, Seq: 2, pl: 29
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Prepared 9 bytes
0c5fce861fb13a9e:     ACK (nb=2), 18, 4, 0

0c5fce861fb13a9e: Sending 54 bytes to 127.0.0.1:4443 at T=0.434696 (5851ccc999fe8)
Select returns 62, from length 16
0c5fce861fb13a9e: Receiving 62 bytes from 127.0.0.1:4443 at T=0.651213 (5851ccc9cedad)
0c5fce861fb13a9e: Receiving packet type: 4 (handshake), S0, Version ff000012,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, <0000000000000002>, Seq: 6, pl: 38
0c5fce861fb13a9e:     Decrypted 21 bytes
0c5fce861fb13a9e:     ACK (nb=0), 0
0c5fce861fb13a9e:     padding, 16 bytes

Processed 62 bytes, state = 13 (client_ready_start), return 0

0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S0,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 3 (3), Phi: 0,
0c5fce861fb13a9e:     Prepared 216 bytes
0c5fce861fb13a9e:     NEW CONNECTION ID[1]: 0xe1fdd8d5bf6ecb39, ae4de4bcec8a0438c8f4cf9c25376d1e
0c5fce861fb13a9e:     NEW CONNECTION ID[2]: 0x0522ee324c648303, 102d4f4fba4e299f516ade6cce61a5b5
0c5fce861fb13a9e:     NEW CONNECTION ID[3]: 0x0ba8d5a218f70d0b, aa71887658f70d6556efa2366a8aa045
0c5fce861fb13a9e:     NEW CONNECTION ID[4]: 0x38026e7774698253, fa9bcace5af252462dc6ba6203e0c67c
0c5fce861fb13a9e:     NEW CONNECTION ID[5]: 0xa2849b98330f73c7, 7f9aa3139736f2a236741a2a54e854b8
0c5fce861fb13a9e:     NEW CONNECTION ID[6]: 0x6edbeccf181b5362, 9567a4c0fd04182862e86599e9dbfbd2
0c5fce861fb13a9e:     NEW CONNECTION ID[7]: 0x547227128e671bc5, ad1b0b39bb79c496df6c7c92069930a3
0c5fce861fb13a9e:     NEW CONNECTION ID[8]: 0xd02d41b90bd79e1d, 8e188c83ade048ee108494ad70fa6f89

0c5fce861fb13a9e: Sending 245 bytes to 127.0.0.1:4443 at T=0.651428 (5851ccc9cee84)
Select returns 65, from length 16
0c5fce861fb13a9e: Receiving 65 bytes from 127.0.0.1:4443 at T=1.078346 (5851ccca3722a)
0c5fce861fb13a9e: Receiving packet type: 6 (1rtt protected), S1,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, Seq: 0 (0), Phi: 0,
0c5fce861fb13a9e:     Decrypted 39 bytes
0c5fce861fb13a9e:     Stream 0, offset 0, length 4, fin = 1: 666f6f0a
0c5fce861fb13a9e:     ACK (nb=0), 0
0c5fce861fb13a9e:     ACK (nb=0), 1
0c5fce861fb13a9e:     ACK (nb=0), 2
0c5fce861fb13a9e:     padding, 16 bytes

Stream 0 ended after 4 bytes
Opening stream 4 to GET /test.html
Opening stream 8 to GET /1234567
Opening stream 12 to GET /main.jpg
Opening stream 16 to GET /war-and-peace.txt
Opening stream 20 to GET /en/latest/
Opening stream 24 to GET /file-123K
Processed 65 bytes, state = 14 (ready), return 0

0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S0,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 4 (4), Phi: 0,
0c5fce861fb13a9e:     Prepared 125 bytes
0c5fce861fb13a9e:     ACK (nb=0), 0
0c5fce861fb13a9e:     Stream 4, offset 0, length 16, fin = 1: 474554202f746573...
0c5fce861fb13a9e:     Stream 8, offset 0, length 14, fin = 1: 474554202f313233...
0c5fce861fb13a9e:     Stream 12, offset 0, length 15, fin = 1: 474554202f6d6169...
0c5fce861fb13a9e:     Stream 16, offset 0, length 24, fin = 1: 474554202f776172...
0c5fce861fb13a9e:     Stream 20, offset 0, length 17, fin = 1: 474554202f656e2f...
0c5fce861fb13a9e:     Stream 24, offset 0, length 16, fin = 1: 474554202f66696c...

0c5fce861fb13a9e: Sending 154 bytes to 127.0.0.1:4443 at T=1.078566 (5851ccca37306)
Select returns 1232, from length 16
0c5fce861fb13a9e: Receiving 1232 bytes from 127.0.0.1:4443 at T=1.265729 (5851ccca64e21)
0c5fce861fb13a9e: Receiving packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, <0000000000000002>, Seq: 36, pl: 1207
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Decrypted 1190 bytes
0c5fce861fb13a9e:     ACK (nb=0), 2
0c5fce861fb13a9e:     padding, 1185 bytes

Processed 1232 bytes, state = 14 (ready), return 0

0c5fce861fb13a9e: Sending packet type: 2 (initial), S0, Version ff000012,
0c5fce861fb13a9e:     <0000000000000002>, <86a99dafa9700ae6>, Seq: 3, pl: 31
0c5fce861fb13a9e:     Token length: 0
0c5fce861fb13a9e:     Prepared 11 bytes
0c5fce861fb13a9e:     ACK (nb=3), 36, 18, 4, 0

0c5fce861fb13a9e: Sending 56 bytes to 127.0.0.1:4443 at T=1.265767 (5851ccca64e47)
Select returns 1232, from length 16
0c5fce861fb13a9e: Receiving 1232 bytes from 127.0.0.1:4443 at T=1.530655 (5851cccaa58ff)
0c5fce861fb13a9e: Receiving packet type: 6 (1rtt protected), S0,
0c5fce861fb13a9e:     <86a99dafa9700ae6>, Seq: 16 (16), Phi: 0,
0c5fce861fb13a9e:     Decrypted 1206 bytes
0c5fce861fb13a9e:     ACK (nb=0), 3
0c5fce861fb13a9e:     padding, 1201 bytes

Processed 1232 bytes, state = 14 (ready), return 0

0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S1,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 5 (5), Phi: 0,
0c5fce861fb13a9e:     Prepared 24 bytes
0c5fce861fb13a9e:     ACK (nb=1), 16, 0
0c5fce861fb13a9e:     padding, 17 bytes

0c5fce861fb13a9e: Sending 53 bytes to 127.0.0.1:4443 at T=1.530685 (5851cccaa591d)
0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S1,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 6 (6), Phi: 0,
0c5fce861fb13a9e:     Prepared 130 bytes
0c5fce861fb13a9e:     Stream 4, offset 0, length 16, fin = 1: 474554202f746573...
0c5fce861fb13a9e:     Stream 8, offset 0, length 14, fin = 1: 474554202f313233...
0c5fce861fb13a9e:     Stream 12, offset 0, length 15, fin = 1: 474554202f6d6169...
0c5fce861fb13a9e:     Stream 16, offset 0, length 24, fin = 1: 474554202f776172...
0c5fce861fb13a9e:     Stream 20, offset 0, length 17, fin = 1: 474554202f656e2f...
0c5fce861fb13a9e:     Stream 24, offset 0, length 16, fin = 1: 474554202f66696c...
0c5fce861fb13a9e:     ACK (nb=1), 16, 0

0c5fce861fb13a9e: Sending 159 bytes to 127.0.0.1:4443 at T=2.952105 (5851cccc00989)
0c5fce861fb13a9e: Sending packet type: 6 (1rtt protected), S1,
0c5fce861fb13a9e:     <0000000000000002>, Seq: 7 (7), Phi: 0,
0c5fce861fb13a9e:     Prepared 130 bytes
0c5fce861fb13a9e:     Stream 4, offset 0, length 16, fin = 1: 474554202f746573...
0c5fce861fb13a9e:     Stream 8, offset 0, length 14, fin = 1: 474554202f313233...
0c5fce861fb13a9e:     Stream 12, offset 0, length 15, fin = 1: 474554202f6d6169...
0c5fce861fb13a9e:     Stream 16, offset 0, length 24, fin = 1: 474554202f776172...
0c5fce861fb13a9e:     Stream 20, offset 0, length 17, fin = 1: 474554202f656e2f...
0c5fce861fb13a9e:     Stream 24, offset 0, length 16, fin = 1: 474554202f66696c...
0c5fce861fb13a9e:     ACK (nb=1), 16, 0



```

**Comment by huitema** [comment]
I found the root cause, coding mistake in the case where server sends ACK when the client has already completed the handshake. But fixing it will require some work, because some reliability tests appear to only work when the extra acks are sent.

**Comment by huitema** [comment]
Fixed in PR #549
